### DISCUSSION TITLE
Samples play when keys are released

### DISCUSSION THREAD

**Discussion by fox632** [marcel-licence/esp32_midi_sampler#61] [Q&A]
Hello.
I built esp32_midi_sampler on esp32 Audio kit v.2.2
Uploaded the sketch, with the AUTO_LOAD_PATCHES_FROM_LITTLEFS parameter
everything worked fine, the wav file was played from memory.
After switching the parameter to AUTO_LOAD_PATCHES_FROM_SD_MMC
Zuki began to play only when the keys were released.

What should I change in the code so that when loading samples from the SD card, the sounds are played when a key is pressed.

Thanks in advance, sorry for my English,
I use google translator for translation.

**Comment by marcel-licence** [comment]
Hello,

I am happy to hear that the general functionality does work. I can imagine that the problem is related to the loop region. There is a function to remove the loop or loop the complete sample. There is also a setting possible resulting in the effect you've observed.

Hope this helps =)

**Reply by fox632** [reply]
Hello.
I figured out the inverse operation of the keys.
I made a small change in the sampler_module file and the keys worked as they should.

```
#ifdef AUTO_LOAD_PATCHES_FROM_LITTLEFS
    if (player->pos - ((float)player->sample_rec->start) >= (player->sample_rec->loop_end))
    {
        uint32_t sampleLenU = sampleLen;
        player->pos -= sampleLenU;
        player->pos_f -= sampleLen - sampleLenU;
    }
#endif
    /* select sd card */
#ifdef AUTO_LOAD_PATCHES_FROM_SD_MMC
    if (player->pos - ((float)player->sample_rec->start) < player->sample_rec->loop_start)
    {
        uint32_t sampleLenU = sampleLen;
        player->pos += sampleLenU;
        player->pos_f += sampleLen - sampleLenU;
    }
#endif
```

It was difficult to figure out why the played sample sounds the same on all keys of the MIDI keyboard.
that is, there is no binding to the notes.

**Reply by marcel-licence** [reply]
I pushed a little update today. When loading samples the loop area was set from sample 0 to sample 0 in case no additional configuration was saved. I hope my change would fix the behaviour.

Loading samples from SD card should use the same algorithm when played back. Thank you for sharing your results.

**Reply by fox632** [reply]
Hello.
I forgot to clarify that I am using esp32 Audio kit v2.2 with esp32-A1S module,
I downloaded the new version of esp32_MIDI_sampler, without changing anything, uploaded it to esp32 Audio kit v2.2,
I checked the work with SD card and flash, now it works in both modes without my addition.
Thank you for your good support.

Now there is one problem left, for some reason, when you press the keys there is no binding to notes, it plays the same on all keys.

To rule out a problem with my equipment.
For the test I decided to use
#ifdef NOTE_ON_AFTER_SETUP
    Sampler_NoteOn(0, 60, 1);
#endif
As a result, when calling the Sampler_NoteOn(0, /* any key code.*/, 1);
with any key code, the sound is played the same way.

If you can also win this problem, it will be generally great.
Huge thanks to you man. =)

**Reply by marcel-licence** [reply]
Thank you for your feedback. 
The first channel is used to play different samples. If you use another MIDI channel you can play samples with different key. I think there will be an update in future for more flexibility.

**Reply by fox632** [reply]
Hello.
I have a simple MIDI keyboard that only has volume controls,
Pitch Bend and Modulation, no other controls.
Therefore, I use only (0) channel, I made a small addition to the code,
or rather in the file: sampler_module.ino
to the Sampler_NoteOnInt function
if (ch == 0)
.....
if (freePlayer != NULL)
.....
freePlayer->normNote = note - 64;//NOTE_NORMAL;
freePlayer->pitch *= pow(2.0f, 1.0f / 12.0f * (note - 64)); /* this would be the a as middle */

After adding these two lines, everything worked on the (0) channel, notes began to play.
I'm putting together a project not for myself, I'm helping a friend, he has a Roland synthesizer, I think everything will work as it should.
Huge thanks to you for your support.
